src/service/document-info-impl.js

```diff
@@ -0,0 +1,91 @@
+/**
+ * Copyright 2015 The AMP HTML Authors. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS-IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import {fromClassForDoc} from '../service';
+import {parseUrl, getSourceUrl} from '../url';
+
+
+/**
+ * Properties:
+ *     - url: The doc's url.
+ *     - sourceUrl: the source url of an amp document.
+ *     - canonicalUrl: The doc's canonical.
+ *     - pageViewId: Id for this page view. Low entropy but should be unique
+ *       for concurrent page views of a user().
+ *
+ * @typedef {{
+ *   url: string,
+ *   sourceUrl: string,
+ *   canonicalUrl: string,
+ *   pageViewId: string,
+ * }}
+ */
+export let DocumentInfoDef;
+
+
+/**
+ * @param {!Node|!./ampdoc-impl.AmpDoc} nodeOrDoc
+ * @return {!DocInfo} Info about the doc
+ */
+export function installDocumentInfoServiceForDoc(nodeOrDoc) {
+  return fromClassForDoc(nodeOrDoc, 'documentInfo', DocInfo);
+}
+
+
+export class DocInfo {
+  /**
+   * @param {!./ampdoc-impl.AmpDoc} ampdoc
+   */
+  constructor(ampdoc) {
+    /** @private @const  */
+    this.ampdoc_ = ampdoc;
+    /** @private {?DocumentInfoDef} */
+    this.info_ = null;
+  }
+
+  /** @return {!DocumentInfoDef} */
+  get() {
+    if (this.info_) {
+      return this.info_;
+    }
+
+    const url = this.ampdoc_.getUrl();
+    const sourceUrl = getSourceUrl(url);
+    const rootNode = this.ampdoc_.getRootNode();
+    let canonicalUrl = rootNode && rootNode.AMP
+        && rootNode.AMP.canonicalUrl;
+    if (!canonicalUrl) {
+      const canonicalTag = rootNode.querySelector('link[rel=canonical]');
+      canonicalUrl = canonicalTag
+          ? parseUrl(canonicalTag.href).href
+          : sourceUrl;
+    }
+    const pageViewId = getPageViewId(this.ampdoc_.win);
+    return this.info_ = {url, sourceUrl, canonicalUrl, pageViewId};
```

Comment:
Per https://github.com/ampproject/amphtml/pull/5606#discussion_r85412412, how about not caching the `url` and `sourceUrl` attributes? This is wayy better than having a getter property.